### PR TITLE
gadget: add kernel command-line remove filter and append to gadget yaml

### PR DESCRIPTION
**Pull request by valentindavid**
TODO:
* [x] Depends on #12979
* [x] Depends on #12976
* [x] Add tests

### CODE REVIEW

**Review by alfonsosanchezbeato — COMMENTED** [review]
I have some minor comments

File: boot/cmdline.go [hunk 1]
```diff
@@ -126,7 +135,10 @@ func bootVarsForTrustedCommandLineFromGadget(gadgetDirOrSnapPath, cmdlineAppend
 	if full {
 		args["snapd_full_cmdline_args"] = extraOrFull
 	} else {
-		args["snapd_full_cmdline_args"] = strutil.JoinNonEmpty([]string{defaultCmdline, extraOrFull}, " ")
+		args["snapd_full_cmdline_args"] = strutil.JoinNonEmpty(append(keepDefaultArgs, extraOrFull), " ")
+	}
+	if len(args["snapd_full_cmdline_args"]) == 0 {
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply boot/cmdline.go:140]: Exactly. It needs a comment for sure!

File: osutil/kcmdline/kcmdline.go [hunk 2]
```diff
@@ -473,3 +473,16 @@ func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) err
 	}
 	return kap.unmarshalFromString(arg)
 }
+
+func FilterOut(cmdline string, match []ArgumentPattern) []string {
```
Comment (alfonsosanchezbeato) [inline osutil/kcmdline/kcmdline.go:477]: This would need a doc string

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for the changes! Minor comment below if you want to address.

File: gadget/gadget.go [hunk 3]
```diff
@@ -1779,37 +1781,59 @@ func isKernelArgumentAllowed(arg string) bool {
...
 	if err != nil {
-		return "", false, fmt.Errorf("cannot open gadget snap: %v", err)
+		return "", false, []kcmdline.ArgumentPattern{}, fmt.Errorf("cannot open gadget snap: %v", err)
 	}
+
+	info, err := ReadInfoFromSnapFileNoValidate(sf, model)
+
+	if err != nil {
+		return "", false, []kcmdline.ArgumentPattern{}, fmt.Errorf("Cannot read snap info: %v", err)
+	}
```
Comment (Meulengracht) [inline gadget/gadget.go:1795]: We don't usually have newlines between function-call and error check
```suggestion
	info, err := ReadInfoFromSnapFileNoValidate(sf, model)
	if err != nil {
		return "", false, []kcmdline.ArgumentPattern{}, fmt.Errorf("Cannot read snap info: %v", err)
	}
```

File: osutil/kcmdline/kcmdline.go [hunk 4]
```diff
@@ -473,3 +473,18 @@ func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) err
 	}
 	return kap.unmarshalFromString(arg)
 }
+
+// FilterOut parses a kernel command line and remove arguments that
+// matches match patterns.
+func FilterOut(cmdline string, match []ArgumentPattern) []string {
```
Comment (Meulengracht) [inline osutil/kcmdline/kcmdline.go:479]: `FilterOut` feels a bit incomplete for a function name. How about `ExcludeMatchingFilter` or `RemoveMatchingFilter` instead?

File: osutil/kcmdline/kcmdline.go [hunk 5]
```diff
@@ -473,3 +473,18 @@ func (kap *ArgumentPattern) UnmarshalYAML(unmarshal func(interface{}) error) err
...
+	parsedDefault := Parse(cmdline)
+	matcher := NewMatcher(match)
+	keepDefaultArgs := []string{}
```
Comment (Meulengracht) [inline osutil/kcmdline/kcmdline.go:482]: This style is preferred for variables that just use their default initializer
```suggestion
	var keepDefaultArgs []string
```

File: gadget/gadget.go [hunk 6]
```diff
@@ -1779,37 +1781,59 @@ func isKernelArgumentAllowed(arg string) bool {
...
+
+	if len(info.KernelCmdline.Append) > 0 || len(info.KernelCmdline.Remove) > 0 {
+		asStr := []string{}
```
Comment (Meulengracht) [inline gadget/gadget.go:1798]: Styling preferred for default initializers
```suggestion
		var asStr []string
```

File: gadget/gadget.go [hunk 7]
```diff
@@ -1779,37 +1781,59 @@ func isKernelArgumentAllowed(arg string) bool {
...
+		for _, cmd := range info.KernelCmdline.Append {
+			value := cmd.String()
+			if !isKernelArgumentAllowed(value) {
+				return "", false, []kcmdline.ArgumentPattern{}, fmt.Errorf("kernel parameter '%s' is not allowed", value)
```
Comment (Meulengracht) [inline gadget/gadget.go:1802]: Is this path tested?

File: boot/seal_test.go [hunk 8]
```diff
@@ -87,6 +87,25 @@ func mockNamedKernelSeedSnap(rev snap.Revision, name string) *seed.Snap {
...
 
 func mockGadgetSeedSnap(c *C, files [][]string) *seed.Snap {
+	mockGadgetYaml := `
+volumes:
+  volumename:
+    bootloader: grub
+`
+	if files == nil {
+		files = [][]string{}
+	}
```
Comment (Meulengracht) [inline boot/seal_test.go:97]: I don't think this is neccessary? An empty or nil should act the same, especially for append()
```suggestion
```

**Review by Meulengracht — COMMENTED** [review]
It's looking super good, a couple of comments

File: gadget/gadget.go (hunk 7 above)
Comment (valentindavid) [reply gadget/gadget.go:1802]: There was actually a bug! Thank you for spotting it.

**Review by Meulengracht — APPROVED** [review]
Thank you for the changes, looks good +1

**Pull request merged**